Add Mount middleware

SimonPersson opened the pull request:
Attempt at #259. 

This middleware takes a mount point and a `Middleware`. When invoked, the mount point is compared with the incoming request, and if they match the underlying middleware is invoked but with the request path rewritten to have the mount point as root.

Ping @cburgdorf

Ryman commented on src/mount.rs:
```diff
@@ -0,0 +1,84 @@
+
+use hyper::uri::RequestUri::AbsolutePath;
+
+pub struct Mount {
```
Should be able to get static dispatch with the following:

```rust
pub struct Mount<M: Middleware + Send + Sync + 'static> {
    mount_point: String,
    middleware: M
}
```

(It may not be required to specify `+Send+Sync+'static` as they are implied by `Middleware` but I forget if that propagates)

Ryman commented on src/mount.rs:
```diff
@@ -0,0 +1,84 @@
+    ///
+    /// server.utilize(
+    ///     Mount::new("/static_files/",
+    ///                StaticFilesHandler::new("/path/to/serve/")
```
May be worth introducing a trait `Mountable` and implement it for `Nickel` so we can have `server.mount(foo, bar)`.

Ryman commented on src/mount.rs:
```diff
@@ -0,0 +1,84 @@
+impl Middleware for Mount {
+    fn invoke<'mw, 'conn>(&'mw self, req: &mut Request<'mw, 'conn>, res: Response<'mw>)
+        -> MiddlewareResult<'mw> {
+        if let AbsolutePath(path) = req.origin.uri.clone() {
```
Would be nice to only do the clone here if the inner condition is true

Ryman replied:
Using `mem::replace` could be even nicer if possible :dart:

Ryman commented on src/mount.rs:
```diff
@@ -0,0 +1,84 @@
+    fn invoke<'mw, 'conn>(&'mw self, req: &mut Request<'mw, 'conn>, res: Response<'mw>)
+        -> MiddlewareResult<'mw> {
+        if let AbsolutePath(path) = req.origin.uri.clone() {
+            if path.len() >= self.mount_point.len()
```
I think this can be `path.starts_with(self.mount_point)`

Ryman commented on src/mount.rs:
```diff
@@ -0,0 +1,84 @@
+
+#[test]
+fn mount_point_rewrites() {
+    assert!(rewrite_mount_point("/") == "/");
```
`assert_eq!` will have better messages on failure here

Ryman commented on examples/mount.rs:
```diff
@@ -0,0 +1,22 @@
+    }));
+
+    server.utilize(Mount::new("/1/2/3", middleware! {
+        "H3110 w0r1d."
```
I'm not sure these first two usages show anything useful to readers, perhaps printing `req.path_without_query()` would make things clearer?

Ryman commented on examples/mount.rs:
```diff
@@ -0,0 +1,22 @@
+
+    server.utilize(Mount::new("/static/files/", StaticFilesHandler::new("examples/assets/")));
+    server.utilize(Mount::new("/static/files/a", middleware! {
+        "No static file called a!"
```
Maybe a comment saying this is fall-through behavior, either way this is more useful than the top two usages :+1: (to me at least)

Ryman commented:
Looks like it could do the job :+1: 

I did take a look at iron's version mentioned in #259 and after looking at the code I'm left wondering a bit if we could abstract the core logic and both benefit from improvements, something to consider for the future I guess.

As for the api, it does contrast a bit with iron's version where they can mount multiple things with the same mount instance, which I guess could potentially have efficiency benefits if mounting multiple things (while being less efficient for single mounts). I'm happy enough with single mounts for now, but I'm wondering if it's a substantial usecase.

Finally, I'm wondering if this is worth isolating in a separate crate? Or do we just hold off until we have infrastructure in place to make that easy (whenever that is)? Again, I'm probably happy enough either way currently.

SimonPersson commented:
@Ryman, I've been through all your comments and adressed everything except the unnecessary copy when the matching branch is not taken. I can't seem to find an elegant way to solve that. If you can, please tell :) My problem is that all solutions I can think of introduces an extra `if`, and I'd like to not introduce an extra branch where it is not really needed.

Ryman commented:
> My problem is that all solutions I can think of introduces an extra if, and I'd like to not introduce an extra branch where it is not really needed.

What's the concern with an extra branch?

Ryman commented on examples/mount.rs:
```diff
@@ -0,0 +1,18 @@
+     * Fall-through behaviour, if StaticFilesHandler does not find a matching file,
+     * the request uri must be reset so that it can be matched against other middleware.
+     */
+    server.mount("/static/files/", StaticFilesHandler::new("examples/assets/"));
```
Looking at this again, the rewrite rules would make this `"/static/files"`? In that case, does this mean that the mount will activate for paths such as `"/static/files_and_stuff/bar"` and pass the uri as `"_and_stuff/bar"`?

I'm also curious about the behavior of `"/"`, as the uri passed to the mounted middleware would be `""`  if the base route is caught (i.e. the client sends `"/"`)? I would have expected `"/"` to be passed?

I'm not really sure what motivates the rewrite rules, can you add a comment to `rewrite_mount_point` with expected behavior?

SimonPersson commented:
>>
My problem is that all solutions I can think of introduces an extra if, and I'd like to not introduce an extra branch where it is not really needed.
>
What's the concern with an extra branch?

No big concerns, just that it looks a bit ugly. If there's a way around it, I'd like to avoid it. Otherwise, no big deal.

>
Looking at this again, the rewrite rules would make this "/static/files"? In that case, does this mean that the mount will activate for paths such as "/static/files_and_stuff/bar" and pass the uri as "_and_stuff/bar"?
>
Im also curious about the behavior of "/", as the uri passed to the mounted middleware would be "" if the base route is caught (i.e. the client sends "/")? I would have expected "/" to be passed?

Good catch, that's probably right! It's not really related to the rewrite rules though, even without those, this same issue would arise. If the mount path is "/static/files", then we also match "/static/files_and_stuff". If the mount path is "/static/files/", then a request to "/static/files/abc" ends up with the path "abc" where I'd expect it to be "/abc". I think this will work if we make sure that mount paths always end with a slash, and then prepend a slash to the path before calling the underlying middleware. That should solve both those failing cases.

>
I'm not really sure what motivates the rewrite rules, can you add a comment to rewrite_mount_point with expected behavior?

I'm not sure about them either. We can also just document which format mount points must be written in and hope that users get it right. Or, we could check for leading and trailing slash and return an error if the user got it wrong.

Ryman commented:
> Or, we could check for leading and trailing slash and return an error if the user got it wrong.

Assuming no-one is trying to mount things on an already running server I think this is a good option which allows people to fail *quickly* at runtime so they can get their stuff working with least surprise. Does make the usage kind of ugly though if it requires an unwrap, would this be a good place to just panic on the basis that it's most likely not what the user intended?

cc @cburgdorf

Ryman commented on src/mount.rs:
```diff
@@ -0,0 +1,95 @@
+impl<M: Middleware> Middleware for Mount<M> {
+    fn invoke<'mw, 'conn>(&'mw self, req: &mut Request<'mw, 'conn>, res: Response<'mw>)
+        -> MiddlewareResult<'mw> {
+        if let AbsolutePath(path) = req.origin.uri.clone() {
```
*untested but might work and isn't totally ugly?*

```rust
let subpath = match req.origin.uri {
    AbsolutePath(ref path) if path.starts_with(self.mount_point) => {
        AbsolutePath(path[self.mount_point.len()..].to_owned())
    },
    _ => return Ok(Continue(res))
};

let original = mem::replace(&mut req.origin.uri, subpath);
let result = self.middleware.invoke(req, res);
req.origin.uri = original;
result
```

SimonPersson replied:
Of course! I never thought to return early in the non mathing case! This will probably work and looks nicer than what I could think of :)

cburgdorf commented:
I've no idea why github doesn't let me add my comment where it belongs so I'll add it here.

>would this be a good place to just panic on the basis that it's most likely not what the user intended?

I'm absolutely in favor of the `panic!` here. It's probably the best way to notify the developer about a misconfiguration.

GitCop commented:
Thanks for contributing! Unfortunately, I'm here to tell you there were style issues with your Pull Request:

 * Commit: a023aa12520a076d1cfcafcb253d58933cd4631b
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: dc8043c022977c4386670f7f85c3a3096b06eb6e
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 9169c1d0cdb30a66d896137fe5bb21d199b56be1
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/nickel-org/nickel.rs/blob/master/contributing.md

---
This message was auto-generated by https://gitcop.com

SimonPersson commented:
I've fixed the broken logic discovered by @Ryman, removed the rewrite rules and replaced them with a `panic!`, and fixed the unnecessary copy when the mount point does not match the request.

homu commented:
:umbrella: The latest upstream changes (presumably #270) made this pull request unmergeable. Please resolve the merge conflicts.

cburgdorf commented:
Yay! We finally have a bot! Love it :)

GitCop commented:
Thanks for contributing! Unfortunately, I'm here to tell you there were style issues with your Pull Request:

 * Commit: a023aa12520a076d1cfcafcb253d58933cd4631b
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: dc8043c022977c4386670f7f85c3a3096b06eb6e
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 9169c1d0cdb30a66d896137fe5bb21d199b56be1
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 3e7bda58147ef241e4d26b106d7e242d332105b1
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 807566e99191d309cf63cac2c7b1969d767a6bb4
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/nickel-org/nickel.rs/blob/master/contributing.md

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Thanks for contributing! Unfortunately, I'm here to tell you there were style issues with your Pull Request:

 * Commit: a023aa12520a076d1cfcafcb253d58933cd4631b
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: dc8043c022977c4386670f7f85c3a3096b06eb6e
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 9169c1d0cdb30a66d896137fe5bb21d199b56be1
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 3e7bda58147ef241e4d26b106d7e242d332105b1
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 807566e99191d309cf63cac2c7b1969d767a6bb4
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 41934476e99ec2e27964f3a9d60a1c8b963a709c
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/nickel-org/nickel.rs/blob/master/contributing.md

---
This message was auto-generated by https://gitcop.com

Ryman commented:
@SimonPersson If you reword commits or squash I think we're good to go with this for now? Or perhaps we should test @cburgdorf's project with commits from this branch?

SimonPersson commented:
I squashed the commits, let's see if the tests pass and you can feel free to merge it!

Ryman commented:
@homu r+

homu commented:
:pushpin: Commit 0000000 has been approved by `Ryman`

<!-- @homu r=Ryman 0000000000000000000000000000000000000000 -->

Ryman commented:
Err, it's definitely not commit 00000000

@homu r-

Ryman commented:
@homu r-

Ryman commented:
@homu r+

homu commented:
:pushpin: Commit 0000000 has been approved by `Ryman`

<!-- @homu r=Ryman 0000000000000000000000000000000000000000 -->

Ryman commented:
Sigh :D

Ryman commented:
@homu r+

homu commented:
:pill: [DEBUG] head_sha reset from 0000000000000000000000000000000000000000 to 0f258361b0ca670321e1f621189a716c94633773

homu commented:
:pushpin: Commit 0f25836 has been approved by `Ryman`

<!-- @homu r=Ryman 0f258361b0ca670321e1f621189a716c94633773 -->

homu commented:
:hourglass: Testing commit 0f25836 with merge 0622106...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/nickel-org/nickel.rs/builds/77834049)

Ryman commented:
Thanks @barosl! :surfer:

cburgdorf commented:
That's awesome! I'll test that with my classroom project later this week.
